packages/discount-code-exporter/src/main.js

```diff
@@ -1,3 +1,165 @@
-export default function findSum (a, b) {
-  return a + b
+/* @flow */
+import type {
+  ApiConfigOptions,
+  CodeData,
+  ImporterOptions,
+  LoggerOptions,
+} from 'types/discountCodes'
+import type {
+  Client,
+  ClientRequest,
+} from 'types/sdk'
+import { createClient } from '@commercetools/sdk-client'
+import { createRequestBuilder } from '@commercetools/api-request-builder'
+import { createHttpMiddleware } from '@commercetools/sdk-middleware-http'
+import {
+  createAuthMiddlewareForClientCredentialsFlow,
+} from '@commercetools/sdk-middleware-auth'
+import {
+  createUserAgentMiddleware,
+} from '@commercetools/sdk-middleware-user-agent'
+import csv from 'fast-csv'
+import JSONStream from 'JSONStream'
+import { flatten } from 'flat'
+import pkg from '../package.json'
+
+export default class DiscountCodeExport {
+  // Set flowtype annotations
+  apiConfig: ApiConfigOptions;
+  batchSize: number;
+  client: Client;
+  logger: LoggerOptions;
+  accessToken: string;
+  delimiter: string
+  exportFormat: string;
+  predicate: string;
+  multiValueDelimiter: string
+  _processCode: Function;
+
+  constructor (
+    options: ImporterOptions,
+    logger: LoggerOptions,
+  ) {
+    if (!options.apiConfig)
+      throw new Error('The contructor must be passed an `apiConfig` object')
+    if (options.batchSize > 500)
+      throw new Error('The `batchSize` must not be more than 500')
+    this.apiConfig = options.apiConfig
+    this.client = createClient({
+      middlewares: [
+        createAuthMiddlewareForClientCredentialsFlow(this.apiConfig),
+        createUserAgentMiddleware({
+          libraryName: pkg.name,
+          libraryVersion: pkg.version,
+        }),
+        createHttpMiddleware({ host: this.apiConfig.apiUrl }),
+      ],
+    })
+
+    this.accessToken = options.accessToken || ''
```

Comment:
I'd refer to a comment from @adnasa:
>should we really mutate the constructor this way?
I would have really preferred that we explicit set it one option at a time instead of assign the whole object.
With this approach, it would be possible for users to do the following
```
const discountCodeExport = new DiscountCodeExport({ run: () => {} /* overrides the instance run method */ })
```